### PR TITLE
Add "View beatmap" button to Discord Rich Presence

### PR DESCRIPTION
**Pull request by dsgpaiva**
This PR adds a button to Discord's RPC so that people can directly check what map a person is playing.
When the user is playing an unpublished / offline map, the button will not exist.

This has been tested, and it works with all gamemodes.

### Online map:
Online map

### Offline map:
Offline map

### CODE REVIEW

**Comment by LittleEndu** [comment]
Something to note for the future - last time I messed with discord API, buttons and secrets (join/spectate) were mutually exclusive. If this is merged we would need to emulate spectating using buttons.

**Comment by jaxxibae** [comment]
One could just scrap that example (the Beatmap Link one) and figure out the way to implement join and spectate on lazer, if we were to follow like it is on stable.

File: osu.Desktop/DiscordRichPresence.cs
Comment (LittleEndu) [inline osu.Desktop/DiscordRichPresence.cs:104]: When lazer starts handling the `osu://` protocol, this should be changed to `osu://b/` and skip opening a web page. Can confirm this works in discord (just right now it opens stable's osu!direct instead of lazer's beatmap listing)

File: osu.Desktop/DiscordRichPresence.cs
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:170]: no reason to return `string`, should return `int?` instead

File: osu.Desktop/DiscordRichPresence.cs
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:175]: please remove the negation from this check

File: osu.Desktop/DiscordRichPresence.cs
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:172]: this is way too nested, should be a single `if` rather than an `if` nested in a `switch`

**Review by bdach — DISMISSED** [review]
this is PR'd from master rather than a feature branch - the contribution guidelines mention this

**Comment by jaxxibae** [comment]
> this is PR'd from master rather than a feature branch - the contribution guidelines mention this

Sure thing, I forgot to read that point. Should I open a new PR from another branch or stick to this one for now?

**Comment by bdach** [comment]
it's fine. just annoying since we won't be able to apply any changes ourselves here if we wanted to, you will have to make all requested changes.

File: osu.Desktop/DiscordRichPresence.cs
Comment (frenzibyte) [reply osu.Desktop/DiscordRichPresence.cs:104]: Also, should probably go with `Open Beatmap` or similar instead:
```suggestion
                        new Button() { Label = "Open Beatmap", Url = $"https://osu.ppy.sh/b/{getOnlineID(activity.Value)}" }
```

**Comment by frenzibyte** [comment]
Just to note out, review conversations must not be resolved by anyone else but the reviewer themself, do not resolve them.

**Comment by bdach** [comment]
Source seems saner now, but I haven't tested this yet and am generally unsure if we want this as a temporary measure rather than wait for the protocol handling to come in and properly add join/spectate buttons via that. Dismissing my review but not approving either on that basis.

File: osu.Desktop/DiscordRichPresence.cs
Comment (peppy) [inline osu.Desktop/DiscordRichPresence.cs:104]: I'm not sure we should be using this link style, as it's a legacy one. In theory the set ID should also be available in all cases the difficulty is, so I think we can form the full URLs here already?

File: osu.Desktop/DiscordRichPresence.cs
Comment (peppy) [inline osu.Desktop/DiscordRichPresence.cs:172]: I'd probably just inline this whole method.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
I've brought this up-to-date, fixed multiple issues with it, added editor support and think it should be fine to go with for now.

**Review by peppy — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Comment by bdach** [comment]
I don't know if it's a linux thing, or a "my system is broken beyond belief" thing, but this doesn't really work for me. As in, it works properly for the very first beatmap I play - the button shows up and the URL is correct. But then when I change beatmaps, all of the other rich presence info updates correctly, the button is still there, but the URL is still pointing at the very first beatmap that I picked in that game run.

Can anyone else verify that this isn't just my own breakage?

**Comment by LittleEndu** [comment]
> Something to note for the future - last time I messed with discord API, buttons and secrets (join/spectate) were mutually exclusive. If this is merged we would need to emulate spectating using buttons.

This is still true, discord just gives you an error `'secrets cannot currently be sent with buttons'` if you have both. If this is merged, the button can only be there in single player. For multiplayer you need secrets, otherwise sending chat invites in discord would not work.

File: osu.Desktop/DiscordRichPresence.cs
Comment (bdach) [inline osu.Desktop/DiscordRichPresence.cs:111]: Interesting curveball here: what should happen if `ruleset.Value` is a custom ruleset and the current beatmap is converted to it? The constructed link will technically not be valid, but it appears that web has some sort of safety against this and will redirect to the correct URL somehow (using beatmap set/beatmap IDs, I presume)?

This *works* and doesn't seem to break, so I'm approving, but I'm not sure it should be allowed, so I'm going to keep away from the merge button for the time being.

**Review by bdach — APPROVED** [review]

**Comment by peppy** [comment]
> This is still true, discord just gives you an error `'secrets cannot currently be sent with buttons'` if you have both. If this is merged, the button can only be there in single player. For multiplayer you need secrets, otherwise sending chat invites in discord would not work.

Am aware, Those buttons would supersede this one. Seems harmless to have this for now (unless you could some users getting potentially annoyed when it changes in the future).

File: osu.Desktop/DiscordRichPresence.cs
Comment (peppy) [reply osu.Desktop/DiscordRichPresence.cs:111]: I'd hope web can support it, although in my testing it didn't seem to perform the difficulty selection correctly with a bad ruleset name.

**Pull request merged**
